torch/csrc/dynamo/guards.cpp

```diff
@@ -574,6 +574,367 @@ static struct PyModuleDef _module = {
     -1,
     _methods};
 
+/**
+ * Stores relevant guard debug information, e.g., failure str for a LeafGuard
+ * failure. The data structure is also accessible in Python.
+ */
+struct GuardDebugInfo {
+  GuardDebugInfo(int num_guards_executed, std::string failure_reason) {
+    this->num_guards_executed = num_guards_executed;
+    this->failure_reason = failure_reason;
+  }
+
+  std::string repr() {
+    return "GuardDebugInfo(num_guards_executed=" +
+        std::to_string(num_guards_executed) +
+        ", failure_reason=" + failure_reason + ")";
+  }
+
+  // Failure reason for a leaf guard.
+  std::string failure_reason;
+
+  // Total number of executed guards so far.
+  int num_guards_executed;
+};
+
+/**
+ * Base class for the terminal or leaf guard in the GuardManager hierarchy.
+ */
+class LeafGuard {
+ public:
+  // Runs the guard and prepares a GuardDebugInfo object.
+  std::pair<bool, GuardDebugInfo> check_with_debug_info(py::object value) {
```

Comment:
Why are you taking the arg by value?

Also, in C++, extremely unusual to do a bool pair. Instead, I would make the object implement the bool operator to represent pass/fail, or do it via static typing.